File: cache-side-channel.js
const FLAG = "Bruin";



// CHALLENGE

const MEMORY_SIZE = 256;
const VICTIM_MEMORY_START = 0;
const VICTIM_MEMORY_END = 64;
const ATTACKER_MEMORY_START = 64;
const ATTACKER_MEMORY_END = 128;

const CACHE_SIZE = 32;


class CacheEntry {
  constructor(tag = null, value = null) {
    this.tag = tag;
    this.value = value;
  }
}


class System {
  #memory;
  #cache;

  constructor() {
    // Initialize memory with random 1-byte values
    this.#memory = new Array(MEMORY_SIZE);
    for (let i = 0; i < MEMORY_SIZE; i++) {
      this.#memory[i] = Math.floor(Math.random() * Math.pow(2, 8));
    }

    // Initialize direct-mapped cache with 1-byte blocks
    this.#cache = new Array(CACHE_SIZE);
    for (let i = 0; i < CACHE_SIZE; i++) {
      this.#cache[i] = new CacheEntry();
    }
  }

  #memoryAccess(address) {
    let start = Date.now();
    while (Date.now() < start + 3000);
    return this.#memory[address];
  }

  #access(address) {
    let index = address % CACHE_SIZE;
    let tag = address >> Math.floor(Math.log2(CACHE_SIZE));
    if (this.#cache[index].tag != tag) {
      this.#cache[index] = new CacheEntry(tag, this.#memoryAccess(address));
    }
    return this.#cache[index].value;
  }

  #victimAccess(address) {
    if (address < VICTIM_MEMORY_START || address >= VICTIM_MEMORY_END) {
      console.log("SEGMENTATION FAULT");
      return;
    }
    return this.#access(address);
  }

  attackerAccess(address) {
    if (address < ATTACKER_MEMORY_START || address >= ATTACKER_MEMORY_END) {
      console.log("SEGMENTATION FAULT");
      return;
    }
    return this.#access(address);
  }

  victimProgram() {
    for (let i = 0; i < FLAG.length; i++) {
      this.#victimAccess(FLAG[i].charCodeAt() - 65);
    }
    console.log("VICTIM PROGRAM COMPLETE");
  }
}

sys = new System();



// SOLUTION

// Prime
for (let address = ATTACKER_MEMORY_START; address < ATTACKER_MEMORY_START + CACHE_SIZE; address++) {
  sys.attackerAccess(address);
}
console.log("PRIME COMPLETE");
console.log("\n");

// Victim
sys.victimProgram();
console.log("\n");

// Probe
for (let address = ATTACKER_MEMORY_START; address < ATTACKER_MEMORY_START + CACHE_SIZE; address++) {
  console.log("Cache Block: " + address % CACHE_SIZE);
  console.log("Corresponding Letter: " + String.fromCharCode(address % CACHE_SIZE + 65));
  let before = Date.now();
  sys.attackerAccess(address);
  console.log("Time: " + Math.floor((Date.now() - before) / 1000) + " seconds");
  console.log("\n");
}
console.log("PROBE COMPLETE");
console.log("\n");

File: cache_side_channel.py
FLAG = "Bruins"



# CHALLENGE

from math import log2
from random import randrange
from time import sleep


MEMORY_SIZE = 256
VICTIM_MEMORY_START = 0
VICTIM_MEMORY_END = 64
ATTACKER_MEMORY_START = 64
ATTACKER_MEMORY_END = 128

CACHE_SIZE = 32


class _cache_entry:
    def __init__(self, tag=None, value=None):
        self.tag = tag
        self.value = value


# Byte-addressable memory
_memory = [randrange(2**8) for _ in range(MEMORY_SIZE)]
def _memory_access(address):
    sleep(3)
    return _memory[address]

# Direct-mapped cache with byte-sized blocks
_cache = [_cache_entry() for _ in range(CACHE_SIZE)]
def _cache_access(address):
    index = address % CACHE_SIZE
    tag = address >> int(log2(CACHE_SIZE))
    if _cache[index].tag != tag:
        _cache[index] = _cache_entry(tag, _memory_access(address))
    return _cache[index].value

def _victim_access(address):
    if address < VICTIM_MEMORY_START or address >= VICTIM_MEMORY_END:
        print("SEGMENTATION FAULT")
        return
    return _cache_access(address)


def attacker_access(address):
    if address < ATTACKER_MEMORY_START or address >= ATTACKER_MEMORY_END:
        print("SEGMENTATION FAULT")
        return
    return _cache_access(address)

def victim_program():
    for char in FLAG:
        address = ord(char) - 65
        value = _victim_access(address)
    print("VICTIM PROGRAM COMPLETE")



# SOLUTION

from time import time


# Prime
for address in range(ATTACKER_MEMORY_START, ATTACKER_MEMORY_START + CACHE_SIZE):
    attacker_access(address)
print("PRIME COMPLETE")
print()

# Victim
victim_program()
print()

# Probe
for address in range(ATTACKER_MEMORY_START, ATTACKER_MEMORY_START + CACHE_SIZE):
    print(f"Cache Block: {address % CACHE_SIZE}")
    print(f"Corresponding Letter: {chr(address % CACHE_SIZE + 65)}")
    before = time()
    attacker_access(address)
    print(f"Time: {int(time() - before)} seconds")
    print()
print("PROBE COMPLETE")
